# Assignment 1

## Problem A (25 points)

__A1.__ In this problem, you will be working with the [Seinfeld Chronicles dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles). Create an account on [Kaggle](https://www.kaggle.com) and download the `scripts.csv` file from the dataset and move it into the `data` directory. Read the `data/scripts.csv` file as a text file line-by-line and examine the list you have loaded the data into. (3 points)

<font color=blue>Here, we read the file line by line and print them:</font>

In [2]:
ContentOfSeinfeld = []
with open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/scripts.csv", "r") as file_handle:
    for line in file_handle:
        ContentOfSeinfeld.append(line)
print(ContentOfSeinfeld[:10])

[',Character,Dialogue,EpisodeNo,SEID,Season\n', '0,JERRY,"Do you know what this is all about? Do you know, why were here? To be out, this is out...and out is one of the single most enjoyable experiences of life. People...did you ever hear people talking about We should go out? This is what theyre talking about...this whole thing, were all out now, no one is home. Not one person here is home, were all out! There are people tryin to find us, they dont know where we are. (on an imaginary phone) Did you ring?, I cant find him. Where did he go? He didnt tell me where he was going. He must have gone out. You wanna go out you get ready, you pick out the clothes, right? You take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...Then youre standing around, whatta you do? You go We gotta be getting back. Once youre out, you wanna get back! You wanna go to sleep, you wanna get up, you wanna go out again tomorrow, right? Where ever you are in life,

__A2.__ Is it possible to work with this data, simply splitting by a delimiter? Explain any complexity in the data's structured format that necessitates an established format-specific file reader. (2 points)

<font color=blue>This data is formated as a list. The elements of the list are seperated with commas. In order to start working with this data we need and structured dataframe which consists of rows and columns. All the columns should be of a certain data type. For example, splitting by a delimiter like comma will lead to splitting the text itself at the comma points, which was not intended. Also, season ID is redundant since episode number and season number columns are existant. Additionally, all the info in one line counts as one single element of the list which we need to seperate into various columns in order to clean the data.</font>

__A3.__ Use the `csv` module to read the contents of the `data/scripts.csv` file into a list. Examine this list. How many unique speaking characters are present in the scripts in total? (5 points)

In [3]:
import csv
import numpy as np
Line_reader = csv.reader(open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/scripts.csv", "r"))
Content = {}
i = 0
for line in Line_reader:
    if i: #newly initiated list is false
        for j in range(len(header)):
            heading = header[j]
            datum = line[j]
            Content[heading].append(datum)
    else:
        header = line
        for heading in header:
            Content[heading] = []
    i += 1
Characters = Content['Character']
type(Characters)
print(Characters[:10])

['JERRY', 'JERRY', 'GEORGE', 'JERRY', 'GEORGE', 'JERRY', 'GEORGE', 'JERRY', 'CLAIRE', 'GEORGE']


In [4]:
myset = set(Characters)
Speakers_Num = len(myset)
print(Speakers_Num)

1639


<font color=blue>Thus, we have 1639 characters talking in the TV series. Although, we can see that this number is not accurate since there are few instances that the speaker name is repeated in another form such as voice-over or for example Jerry(shouting) can be a seperate character here which is not accurate.</font>

__A4.__ Count the dialogue entries for the four major characters, "JERRY", "GEORGE", "ELAINE", and "KRAMER", using a dictionary (you are not allowed to use the Counter data structure for any component of this problem). (2 points) 

#### Number of dialogue enteries for JERRY:

In [5]:
len([i for i in Characters if i == "JERRY"])

14786

#### Number of dialogue enteries for GEORGE:

In [6]:
len([i for i in Characters if i == "GEORGE"])

9708

#### Number of dialogue enteries for ELAINE:

In [7]:
len([i for i in Characters if i == "ELAINE"])

7983

#### Number of dialogue enteries for KRAMER:

In [8]:
len([i for i in Characters if i == "KRAMER"])

6664

__A5.__ Count the number of words spoken by each of the main characters using a dictionary. (3 points)

<font color=blue>First, I searched for each characters Speech lines seperatly and then we can count the words for each of them.</font>

#### Word count for JERRY:

In [9]:
JerrySpeech = [Content['Dialogue'][i] for i in range(len(Content['Character'])) if Content['Character'][i] == "JERRY"]

In [10]:
num_wordsjerry = [len(sentence.split()) for sentence in JerrySpeech]
jerry_words = sum(num_wordsjerry)
jerry_words

147389

#### Word count for GEORGE:

In [11]:
GeorgeSpeech = [Content['Dialogue'][i] for i in range(len(Content['Character'])) if Content['Character'][i] == "GEORGE"]

In [12]:
num_wordsgeorge = [len(sentence.split()) for sentence in GeorgeSpeech]
george_words = sum(num_wordsgeorge)
george_words

107029

#### Word count for ELAINE:

In [13]:
ElaineSpeech = [Content['Dialogue'][i] for i in range(len(Content['Character'])) if Content['Character'][i] == "ELAINE"]

In [14]:
num_wordselaine = [len(sentence.split()) for sentence in ElaineSpeech]
elaine_words = sum(num_wordselaine)
elaine_words

74634

#### Word count for KRAMER:

In [15]:
KramerSpeech = [Content['Dialogue'][i] for i in range(len(Content['Character'])) if Content['Character'][i] == "KRAMER"]

In [16]:
num_wordskramer = [len(sentence.split()) for sentence in KramerSpeech]
kramer_words = sum(num_wordskramer)
kramer_words

70299

__A6.__ Count how many times each word is spoken by the main characters using a dictionary, then sort these word counts in descending order, i.e. from the most commonly spoken words to least. [__Hint__: You can use either a lambda function or a list comprehension to do this.] (5 points)

#### Most spoken words of Jerry:

In [17]:
import re
JerryDict = {}
for sentence in JerrySpeech:
    for word in re.split('\s', sentence): # split with whitespace
        try:
            JerryDict[word] += 1
        except KeyError:
            JerryDict[word] = 1
JerryDict = sorted(JerryDict.items(), key=lambda x: x[1], reverse=True)
print(JerryDict[:50])

[('the', 4708), ('I', 4665), ('you', 3622), ('a', 3294), ('to', 3059), ('of', 1518), ('in', 1304), ('and', 1282), ('You', 1188), ('is', 1181), ('that', 1150), ('it', 1060), ('with', 927), ('on', 904), ("I'm", 884), ("don't", 832), ('have', 803), ('What', 799), ('this', 794), ('was', 793), ('my', 772), ('like', 751), ('do', 749), ('for', 746), ('just', 704), ('know', 693), ('what', 682), ('get', 682), ('not', 672), ('your', 620), ('be', 619), ('me', 596), ('Oh,', 586), ('are', 583), ('about', 571), ('at', 565), ('he', 560), ('Well,', 557), ('out', 554), ('it.', 517), ('got', 502), ('think', 473), ('we', 461), ('his', 451), ('up', 451), ('Yeah,', 450), ("it's", 445), ('So', 426), ('all', 412), ('they', 410)]


#### Most spoken words by George:

In [18]:
GeorgeDict = {}
for sentence in GeorgeSpeech:
    for word in re.split('\s', sentence): # split with whitespace
        try:
            GeorgeDict[word] += 1
        except KeyError:
            GeorgeDict[word] = 1
GeorgeDict = sorted(GeorgeDict.items(), key=lambda x: x[1], reverse=True)
print(GeorgeDict[:50])

[('I', 3946), ('the', 3136), ('a', 2457), ('to', 2287), ('you', 2269), ('of', 1033), ('in', 923), ('and', 922), ('You', 879), ("I'm", 823), ('is', 809), ('it', 786), ('that', 763), ('my', 706), ('have', 646), ('was', 644), ('with', 631), ('for', 605), ('me', 600), ('this', 573), ('like', 543), ('not', 532), ('What', 520), ('do', 518), ('on', 515), ('just', 497), ("don't", 485), ('know', 479), ('be', 471), ('get', 460), ('what', 441), ("It's", 415), ('about', 404), ('are', 383), ('at', 378), ('all', 371), ('think', 371), ('got', 367), ('she', 361), ('Oh,', 339), ('out', 339), ('Well,', 328), ('it.', 326), ('your', 324), ('he', 323), ('her', 312), ('we', 306), ('his', 301), ('up', 298), ("it's", 289)]


#### Most spoken words by Elaine:

In [19]:
ElaineDict = {}
for sentence in ElaineSpeech:
    for word in re.split('\s', sentence): # split with whitespace
        try:
            ElaineDict[word] += 1
        except KeyError:
            ElaineDict[word] = 1
ElaineDict = sorted(ElaineDict.items(), key=lambda x: x[1], reverse=True)
print(ElaineDict[:50])

[('I', 2604), ('the', 1797), ('you', 1713), ('a', 1557), ('to', 1538), ('and', 748), ('is', 711), ('of', 634), ('that', 603), ('You', 598), ('in', 588), ("I'm", 495), ('it', 493), ('Oh,', 490), ('this', 464), ('have', 458), ('just', 444), ('my', 438), ('he', 429), ('was', 416), ('What', 398), ("don't", 384), ('with', 376), ('for', 368), ('me', 360), ('on', 358), ('do', 334), ('know', 329), ('are', 308), ('at', 308), ('not', 305), ('Well,', 303), ('what', 290), ('like', 282), ('be', 275), ('get', 269), ('her', 265), ('got', 261), ('your', 259), ('about', 258), ('Yeah,', 246), ('think', 245), ('so', 240), ('we', 238), ('out', 230), ("It's", 220), ('up', 216), ("it's", 209), ('No,', 200), ('it.', 196)]


#### Most spoken words by Kramer:

In [20]:
KramerDict = {}
for sentence in KramerSpeech:
    for word in re.split('\s', sentence): # split with whitespace
        try:
            KramerDict[word] += 1
        except KeyError:
            KramerDict[word] = 1
KramerDict = sorted(KramerDict.items(), key=lambda x: x[1], reverse=True)
print(KramerDict[:50])

[('the', 2208), ('I', 2155), ('a', 1560), ('you', 1534), ('to', 1382), ('and', 797), ('of', 685), ('in', 607), ("I'm", 551), ('You', 516), ('Well,', 512), ('it', 501), ('that', 498), ('on', 452), ('is', 424), ('my', 415), ('Yeah,', 396), ('Oh,', 391), ('got', 384), ('for', 375), ('this', 355), ('with', 343), ('at', 317), ('was', 313), ('get', 313), ("don't", 309), ('me', 302), ('your', 291), ('know', 290), ('have', 285), ('be', 283), ('like', 278), ('just', 267), ('what', 263), ('out', 259), ('his', 256), ('do', 255), ('gonna', 250), ('are', 246), ('he', 245), ('all', 240), ("It's", 239), ('up', 227), ('Hey,', 226), ('it.', 225), ("it's", 220), ('Jerry,', 217), ('we', 201), ('not', 197), ('know,', 195)]


__A7.__ Load the `data/stop-words.txt` file into a list. Find the 10 most common words for each of the main characters that are not in this list of stop words. Put these most common words in a dictionary data strucutre. (5 points)

#### Reading the text file and making a dictionary:

In [21]:
stop_words = []
with open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/stop-words.txt", "r") as file_handle:
    for line in file_handle:
        stop_words.extend([i for i in line.split()])
print(stop_words)

['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'couldn', "couldn't", 'd', 'did', 'didn', "didn't", 'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven', "haven't", 'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 'isn', "isn't", 'it', "it's", 'its', 'itself', 'just', 'll', 'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself', 'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', "shan't", 'she', "she's", 'should', "should've", 'shouldn', "shouldn't", 'so', 'some',

<font color=blue>Here, we find the top 10 used words for each character that were not in  the list.</font>

#### Jerry's top words:

In [22]:
WordsOfJerry = [JerryDict[i][0].lower() for i in range(len(JerryDict))]
#print(WordsOfJerry)

In [23]:
ValidWordsJerry = [x for x in WordsOfJerry if x not in stop_words]

In [24]:
#Jerry
Top10Jerry = ValidWordsJerry[:10]
print(Top10Jerry)

["i'm", 'like', 'know', 'get', 'oh,', 'well,', 'it.', 'got', 'think', 'yeah,']


#### George's top words:

In [25]:
WordsOfGeorge = [GeorgeDict[i][0].lower() for i in range(len(GeorgeDict))]
#print(WordsOfGeorge)

In [26]:
ValidWordsGeorge = [x for x in WordsOfGeorge if x not in stop_words]

In [27]:
#George
Top10George = ValidWordsGeorge[:10]
print(Top10George)

["i'm", 'like', 'know', 'get', 'think', 'got', 'oh,', 'well,', 'it.', 'see']


#### Elaine's top words:

In [28]:
WordsOfElaine = [ElaineDict[i][0].lower() for i in range(len(ElaineDict))]
#print(WordsOfElaine)

In [29]:
ValidWordsElaine = [x for x in WordsOfElaine if x not in stop_words]

In [30]:
#Elaine
Top10Elaine = ValidWordsElaine[:10]
print(Top10Elaine)

["i'm", 'oh,', 'know', 'well,', 'like', 'get', 'got', 'yeah,', 'think', 'no,']


#### Kramer's top words:

In [31]:
WordsOfKramer = [KramerDict[i][0].lower() for i in range(len(KramerDict))]
#print(WordsOfKramer)

In [32]:
ValidWordsKramer = [x for x in WordsOfKramer if x not in stop_words]

In [33]:
#Kramer
Top10Kramer = ValidWordsKramer[:10]
print(Top10Kramer)

["i'm", 'well,', 'yeah,', 'oh,', 'got', 'get', 'know', 'like', 'gonna', 'hey,']


## Problem B (25 points)

__B1.__ You will be using part of the [Goodbooks 10k dataset](https://github.com/zygmuntz/goodbooks-10k) for this problem. Read the `data/goodreads-books.csv` file into a list. Create a dictionary for each book in the list that contains these fields: authors, original title, original publication year, average rating, and ratings count. (You should convert average rating and ratings count into `float` and `int` types.) Put all these metadata dictionaries into a list. (5 points)

In [34]:
LineOfBooks = csv.reader(open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/goodreads-books.csv", "r"))
Books = {}
i = 0
for line in LineOfBooks:
    if i: #newly initiated list is false
        for j in range(len(header)):
            heading = header[j]
            datum = line[j]
            Books[heading].append(datum)
    else:
        header = line
        for heading in header:
            Books[heading] = []
    i += 1
Books.keys()

dict_keys(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id', 'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year', 'original_title', 'title', 'language_code', 'average_rating', 'ratings_count', 'work_ratings_count', 'work_text_reviews_count', 'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5', 'image_url', 'small_image_url'])

In [35]:
Authors = Books['authors']
OriginalTitle = Books['original_title']
OriginalPublicationYear = Books['original_publication_year']
RatingsCounttemp = Books['ratings_count']
RatingsCount = [int(i) for i in RatingsCounttemp]
AverageRatingtemp = Books['average_rating']
AverageRating = [float(i) for i in AverageRatingtemp]

In [36]:
All_the_book = []

for i in range (len(OriginalTitle)):
    All_the_book.append(dict([("original_title", Books["original_title"][i])]))

for i in range (len(OriginalTitle)):
    All_the_book[i]["authors"] = Authors[i]
    All_the_book[i]["original_publication_year"] = OriginalPublicationYear[i]
    All_the_book[i]["ratings_count"] = RatingsCount[i]
    All_the_book[i]["average_rating"] = AverageRating[i]

print(All_the_book[:10])

[{'original_title': 'The Hunger Games', 'authors': 'Suzanne Collins', 'original_publication_year': '2008.0', 'ratings_count': 4780653, 'average_rating': 4.34}, {'original_title': "Harry Potter and the Philosopher's Stone", 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '1997.0', 'ratings_count': 4602479, 'average_rating': 4.44}, {'original_title': 'Twilight', 'authors': 'Stephenie Meyer', 'original_publication_year': '2005.0', 'ratings_count': 3866839, 'average_rating': 3.57}, {'original_title': 'To Kill a Mockingbird', 'authors': 'Harper Lee', 'original_publication_year': '1960.0', 'ratings_count': 3198671, 'average_rating': 4.25}, {'original_title': 'The Great Gatsby', 'authors': 'F. Scott Fitzgerald', 'original_publication_year': '1925.0', 'ratings_count': 2683664, 'average_rating': 3.89}, {'original_title': 'The Fault in Our Stars', 'authors': 'John Green', 'original_publication_year': '2012.0', 'ratings_count': 2346404, 'average_rating': 4.26}, {'original_t

__B2.__ Write a function to sort this list of book metadata in descending order of average rating. The function should take the list of metadata dictionaries as an input argument. (5 points)

In [37]:
def SortingBooks(metadata):
    SortedList = sorted(metadata, key=lambda k: k['average_rating'], reverse=True)
    return SortedList

In [38]:
All_the_books_sorted = SortingBooks(All_the_book)

__B3.__ Update the function to take two arguments: the list, and an integer value for minimum ratings count. The function should now sort _and_ filter the list, returning a list of books sorted by average rating that have been rated by more than a specified number of users. You can use three different approaches: loops, comprehensions, and the built-in `filter()` function (look up documentation and examples). (10 points)

In [39]:
def FilteredSortingBooks(metadata, n):
    FilteredList = list(filter(lambda k: k['ratings_count'] > n, metadata))
    SortedList = sorted(FilteredList, key=lambda k: k['average_rating'], reverse=True)
    return SortedList

In [40]:
FilteredSortedAllTheBooks = FilteredSortingBooks(All_the_book, 200000)
print(FilteredSortedAllTheBooks[:10])

[{'original_title': 'Harry Potter and the Deathly Hallows', 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '2007.0', 'ratings_count': 1746574, 'average_rating': 4.61}, {'original_title': "The Wise Man's Fear", 'authors': 'Patrick Rothfuss', 'original_publication_year': '2011.0', 'ratings_count': 245686, 'average_rating': 4.57}, {'original_title': 'The Name of the Wind', 'authors': 'Patrick Rothfuss', 'original_publication_year': '2007.0', 'ratings_count': 400101, 'average_rating': 4.55}, {'original_title': 'Harry Potter and the Half-Blood Prince', 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '2005.0', 'ratings_count': 1678823, 'average_rating': 4.54}, {'original_title': 'A Storm of Swords', 'authors': 'George R.R. Martin', 'original_publication_year': '2000.0', 'ratings_count': 469022, 'average_rating': 4.54}, {'original_title': 'The Nightingale', 'authors': 'Kristin Hannah', 'original_publication_year': '2015.0', 'ratings_count': 25360

__Bonus:__ Use all three approaches to write three different versions of the function. See if you can condense your code into a single line for some of the approaches! (10 points)

#### Approach 1: Loops

In [41]:
def LoopsSortingBooks(metadata, n):
    QualifiedBooks = []
    for EachBook in metadata:
        if EachBook['ratings_count'] > n:
            QualifiedBooks.append(EachBook)
    SortedList = sorted(QualifiedBooks, key=lambda k: k['average_rating'], reverse=True)
    return SortedList

In [42]:
LoopsSortedAllTheBooks = LoopsSortingBooks(All_the_book, 200000)
print(LoopsSortedAllTheBooks[:2])

[{'original_title': 'Harry Potter and the Deathly Hallows', 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '2007.0', 'ratings_count': 1746574, 'average_rating': 4.61}, {'original_title': "The Wise Man's Fear", 'authors': 'Patrick Rothfuss', 'original_publication_year': '2011.0', 'ratings_count': 245686, 'average_rating': 4.57}]


#### Approach 2: Comprehensions

In [47]:
def ComprehensionSortingBooks(metadata, n):
    QualifiedBooks = [EachBook for EachBook in metadata if EachBook['ratings_count'] > n]
    SortedList = sorted(QualifiedBooks, key=lambda k: k['average_rating'], reverse=True)
    return SortedList

In [48]:
ComprehensionSortedAllTheBooks = ComprehensionSortingBooks(All_the_book, 200000)
print(ComprehensionSortedAllTheBooks[:2])

[{'original_title': 'Harry Potter and the Deathly Hallows', 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '2007.0', 'ratings_count': 1746574, 'average_rating': 4.61}, {'original_title': "The Wise Man's Fear", 'authors': 'Patrick Rothfuss', 'original_publication_year': '2011.0', 'ratings_count': 245686, 'average_rating': 4.57}]


#### Approach 3: filter()

In [50]:
def FilteredSortingBooks(metadata, n):
    SortedList = sorted(
        list(filter(lambda k: k['ratings_count'] > n, metadata)), key=lambda k: k['average_rating'], reverse=True)
    return SortedList

In [51]:
FilteredSortedAllTheBooks = FilteredSortingBooks(All_the_book, 200000)
print(FilteredSortedAllTheBooks[:2])

[{'original_title': 'Harry Potter and the Deathly Hallows', 'authors': 'J.K. Rowling, Mary GrandPré', 'original_publication_year': '2007.0', 'ratings_count': 1746574, 'average_rating': 4.61}, {'original_title': "The Wise Man's Fear", 'authors': 'Patrick Rothfuss', 'original_publication_year': '2011.0', 'ratings_count': 245686, 'average_rating': 4.57}]


__B3.__ Why is using a function for this task prudent? What do you think is an acceptable minimum ratings count? (5 points)

<font color=blue>It is prudent to use a function with n as the argument for the minimum rating count because it will exclude the titles with very few rating counts which usually has unrealistic ratings. In another word, the average rating is valid with enough data points. For high confidence levels, the rating count should be higher. For example, for 95% confidence level(1.96 is the coefficient for 95% confidence level) and the marginal error being less than 100 books, for this data set, where we have 10000 book titles, the minimum acceptable rating count is:</font>

$$n \geq (1.96(10000)/100)^2 = 38416 $$



## Problem C (50 points)

This problem deals with finding "pangrams" in text. A pangram is a sentence containing all 26 letters of the alphabet. `x` and `y` in the cell below are example sentences, `x` is a pangram, `y` is not.

In [52]:
x = "Jim quickly realized that the beautiful gowns are expensive."
y = "This sentence is most certainly not a pangram."

__C1.__ Define a generator function, `indices()`, that takes a string as input and outputs the index numbers where a letter occurs for the first time in the string. [__Hint:__ you can compare letters like numbers. For example, `char >= "a"` is a valid conditional statement. You can use this to check whether characters in a string are letters.] (5 points)

<font color=blue>When I solved this problem I was not aware of the .index method, so I did not used that in here.</font>

In [53]:
def indices(input_string):
    input_string = input_string.lower()
    Alphabet = "abcdefghijklmnopqrstuvwxyz"
    IndexList = []
    for i in Alphabet:
        checkforletter = [pos for pos, char in enumerate(input_string) if char == i]
        if checkforletter:
            IndexList.append(min(checkforletter))
        else:
            IndexList.append("NA")
        Index_number = IndexList
    yield Index_number

In [54]:
[n for n in indices("My name is kiana")][0]

[4,
 'NA',
 'NA',
 'NA',
 6,
 'NA',
 'NA',
 'NA',
 8,
 'NA',
 11,
 'NA',
 0,
 3,
 'NA',
 'NA',
 'NA',
 'NA',
 9,
 'NA',
 'NA',
 'NA',
 'NA',
 'NA',
 1,
 'NA']

__C2.__ Define a function, `verify()`, that takes a string as input and uses the `indices()` function to check if the string is a pangram. The output should be boolean `True` or `False`. (5 points)

In [55]:
def verify(String_entry):
    String_entry = String_entry.lower()
    Index_number = [n for n in indices(String_entry)][0]
    if "NA" in Index_number:
        return False
    else:
        return True

In [56]:
verify("the quick brown fox jumps over the lazy dog")

True

In [57]:
verify("This sentence is most certainly not a pangram")

False

__C3:__ Write a version of `verify()` named `tiny_verify()` that performs the check in a single line of code, without using `indices()`. [__Hint:__ Use a comprehension.] (5 points)

In [58]:
def tiny_verify(inputString):
    return not set('abcdefghijklmnopqrstuvwxyz') - set(inputString.lower())

In [59]:
tiny_verify("the quick brown fox")

False

__C4.__ Modify the `verify()` function to figure out which letters (if any) are missing from a purported pangram. This version should return the list of missing letters instead of a boolean value. [__Hint:__ You can get a string containing all the letters of the alphabet by importing `ascii_lowercase` from the `string` module.] (5 points)

In [60]:
def verify2(inputString):
    inpuString = inputString.lower()
    Alphabet = "abcdefghijklmnopqrstuvwxyz"
    Index_number = [n for n in indices(inputString)][0]
    ListOfMissingLetters = []
    for i in range(len(Alphabet)):
        if Index_number[i] == "NA":
            ListOfMissingLetters.append(Alphabet[i])  
    return ListOfMissingLetters

In [61]:
verify2("My name is kiana")

['b',
 'c',
 'd',
 'f',
 'g',
 'h',
 'j',
 'l',
 'o',
 'p',
 'q',
 'r',
 't',
 'u',
 'v',
 'w',
 'x',
 'z']

In [62]:
verify2("the quick brown fox jumps over the lazy dog")

[]

__C5.__ Load and iterated through the collected [list of pangrams](http://clagnut.com/blog/2380/) in `data/pangrams.txt` line by line and determine if they are actually pangrams. Print out any lines that are not actually pangrams, and also the letters that are missing. (5 points)

In [63]:
FailedPangrams = []
with open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/pangrams.txt", "r") as file_handle:
    for line in file_handle:
        if verify(line) == False:
            FailedPangrams.append(line)
            print(line)
            print(verify2(line))

Show mangled quartz flip vibe exactly.

['j', 'k']
Unamazingly, this six-word pangram is questionable!

['c', 'f', 'j', 'k', 'v']


__C6:__ Use the output from the `verify()` function to fix the failed pangrams, and verify that you have fixed them. (5 points) 

In [64]:
FixedPangrams = []
for line in FailedPangrams:
    for missingchar in verify2(line):
        line = line + missingchar
    FixedPangrams.append(line)

In [72]:
for element in FixedPangrams:
    print(element, "together is a", verify(element), "pangram")

Show mangled quartz flip vibe exactly.
jk together is a True pangram
Unamazingly, this six-word pangram is questionable!
cfjkv together is a True pangram


__C7.__ In the cell below are provided some information about a set of books. Create a data object that holds the book numbers and titles associated to each authors's name. Write this out as a JSON file in the `data/books/` directory using the following schema. (5 points)

`
books = {
    AuthorName: {
        BookNumber: BookTitle,
        ...
    },
    ...
}
`

In [73]:
# 84.txt; Frankenstein, or the Modern Prometheus; Mary Wollstonecraft (Godwin) Shelley
# 98.txt; A Tale of Two Cities; Charles Dickens 
# 161.txt; Sense and Sensibility; Jane Austen
# 730.txt; Oliver Twist or the Parish Boy's Progress; Charles Dickens
# 768.txt; Wuthering Heights; Emily Brontë
# 1322.txt; Leaves of Grass; Walt Whitman
# 1342.txt; Pride and Prejudice; Jane Austen
# 1400.txt; Great Expectations; Charles Dickens
# 2701.txt; Moby Dick; or the Whale; Herman Melville
# 4300.txt; Ulysses; James Joyce

In [74]:
import json
books = {}  
books['Mary Wollstonecraft (Godwin) Shelley'] = []  
books['Mary Wollstonecraft (Godwin) Shelley'].append({  
    'BookNumber': '84',
    'BookTitle': 'Frankenstein, or the Modern Prometheus'
})
books['Charles Dickens'] = []
books['Charles Dickens'].append({  
    'BookNumber': '98',
    'BookTitle': 'A Tale of Two Cities'
})
books['Charles Dickens'].append({  
    'BookNumber': '730',
    'BookTitle': "Oliver Twist or the Parish Boy's Progress"
})
books['Charles Dickens'].append({  
    'BookNumber': '1400',
    'BookTitle': "Great Expectations"
})
books['Jane Austen'] = []
books['Jane Austen'].append({  
    'BookNumber': '161',
    'BookTitle': "Sense and Sensibility"
})
books['Jane Austen'].append({  
    'BookNumber': '1342',
    'BookTitle': "Pride and Prejudice"
})
books['Emily Brontë'] = []
books['Jane Austen'].append({  
    'BookNumber': '768',
    'BookTitle': "Wuthering Heights"
})
books['Walt Whitman'] = []
books['Walt Whitman'].append({  
    'BookNumber': '1322',
    'BookTitle': "Leaves of Grass"
})
books['Herman Melville'] = []
books['Herman Melville'].append({  
    'BookNumber': '2701',
    'BookTitle': "Moby Dick, or the Whale"
})
books['James Joyce'] = []
books['James Joyce'].append({  
    'BookNumber': '4300',
    'BookTitle': "Ulysses"
})
with open('/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/books/BooksInfo.txt', 'w'
         ) as outfile:  
    json.dump(books, outfile)    

__C8.__ Write a function, `get_pangrams()`, that takes a book number and outputs a list of the book's pangram sentences and the total number of sentences in the book. You will need to use the `re` (regular expressions) module to split the book text into sentences using the `re.split(pattern, string)` function. The pattern you will need is `"[\.\?\!][^a-zA-Z]"`. (5 points)

<font color=blue>In this part we are seperating the sentences in each book roughly. It means that we are ignoring the error caused by more comlex sentences which need some data cleaning.</font>

In [75]:
import re
def get_pangrams(Book_Number):
    address = "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/books/"
    address = address + str(Book_Number) + ".txt"
    pangram = []
    sentences = []
    with open(address, "r") as file_handle:
        allthetext = file_handle.read()
    sentenceswithnewline = re.split("[\.\?\!][^a-zA-Z]", allthetext)
    for i in sentenceswithnewline:
        sentences.append(i.strip('\n'))
    for elements in sentences:
        pangram.append(verify(str(elements)))
    NumberOfPangrams = sum(pangram)
    return([NumberOfPangrams, len(sentences)])

__C9.__ Determine who is the pangrammiest author and what the pangrammiest book is, as determined by pangrams per sentence. [__Hint:__ Use `defaultdict`s to create "pangrams by author" and "pangrams by book" objects.] (10 points)

In [76]:
from collections import defaultdict

#### Calculating the pangrams per sentence for each book:

In [77]:
authorsandpangrams = [("Mary Wollstonecraft (Godwin) Shelley", get_pangrams(84)[0]/get_pangrams(84)[1]), 
        ("Charles Dickens", get_pangrams(98)[0]/get_pangrams(98)[1]), 
        ("Jane Austen", get_pangrams(161)[0]/get_pangrams(161)[1]), 
        ("Charles Dickens", get_pangrams(730)[0]/get_pangrams(730)[1]), 
        ("Emily Brontë", get_pangrams(768)[0]/get_pangrams(768)[1]), 
        ("Walt Whitman", get_pangrams(1322)[0]/get_pangrams(1322)[1]), 
        ("Jane Austen", get_pangrams(1342)[0]/get_pangrams(1342)[1]), 
        ("Charles Dickens", get_pangrams(1400)[0]/get_pangrams(1400)[1]), 
        ("Herman Melville", get_pangrams(2701)[0]/get_pangrams(2701)[1]), 
        ("James Joyce", get_pangrams(4300)[0]/get_pangrams(4300)[1])]

#### Calculating the pangrams per sentence for each author:

In [78]:
pangrams_by_author = defaultdict(list)

for authors, authorsandpangrams in authorsandpangrams:
    pangrams_by_author[authors].append(authorsandpangrams)    
print(pangrams_by_author)

defaultdict(<class 'list'>, {'Mary Wollstonecraft (Godwin) Shelley': [0.0], 'Charles Dickens': [0.00011811953697141507, 9.492168960607498e-05, 0.0], 'Jane Austen': [0.0, 0.00042005040604872583], 'Emily Brontë': [0.00013795006207752792], 'Walt Whitman': [0.00233160621761658], 'Herman Melville': [0.00038387715930902113], 'James Joyce': [0.00044682752457551384]})


In [84]:
summed = {k: sum(v) for (k, v) in pangrams_by_author.items()}
print(summed)

{'Mary Wollstonecraft (Godwin) Shelley': 0.0, 'Charles Dickens': 0.00021304122657749006, 'Jane Austen': 0.00042005040604872583, 'Emily Brontë': 0.00013795006207752792, 'Walt Whitman': 0.00233160621761658, 'Herman Melville': 0.00038387715930902113, 'James Joyce': 0.00044682752457551384}


#### Sorting the pangramiest authors from the most to the least:

In [85]:
for w in sorted(summed, key=summed.get, reverse=True):
  print(w, summed[w])

Walt Whitman 0.00233160621761658
James Joyce 0.00044682752457551384
Jane Austen 0.00042005040604872583
Herman Melville 0.00038387715930902113
Charles Dickens 0.00021304122657749006
Emily Brontë 0.00013795006207752792
Mary Wollstonecraft (Godwin) Shelley 0.0


#### Calculating the pangrams per sentence for each book:

In [86]:
booksandpangrams = [("Frankenstein, or the Modern Prometheus", get_pangrams(84)[0]/get_pangrams(84)[1]), 
        ("A Tale of Two Cities", get_pangrams(98)[0]/get_pangrams(98)[1]), 
        ("Sense and Sensibility", get_pangrams(161)[0]/get_pangrams(161)[1]), 
        ("Oliver Twist or the Parish Boy's Progress", get_pangrams(730)[0]/get_pangrams(730)[1]), 
        ("Wuthering Heights", get_pangrams(768)[0]/get_pangrams(768)[1]), 
        ("Leaves of Grass", get_pangrams(1322)[0]/get_pangrams(1322)[1]), 
        ("Pride and Prejudice", get_pangrams(1342)[0]/get_pangrams(1342)[1]), 
        ("Great Expectations", get_pangrams(1400)[0]/get_pangrams(1400)[1]), 
        ("Moby Dick, or the Whale", get_pangrams(2701)[0]/get_pangrams(2701)[1]), 
        ("Ulysses", get_pangrams(4300)[0]/get_pangrams(4300)[1])]

In [87]:
pangrams_by_book = defaultdict(list)

for title, booksandpangrams in booksandpangrams:
    pangrams_by_book[title].append(booksandpangrams)
    
print(pangrams_by_book)

defaultdict(<class 'list'>, {'Frankenstein, or the Modern Prometheus': [0.0], 'A Tale of Two Cities': [0.00011811953697141507], 'Sense and Sensibility': [0.0], "Oliver Twist or the Parish Boy's Progress": [9.492168960607498e-05], 'Wuthering Heights': [0.00013795006207752792], 'Leaves of Grass': [0.00233160621761658], 'Pride and Prejudice': [0.00042005040604872583], 'Great Expectations': [0.0], 'Moby Dick, or the Whale': [0.00038387715930902113], 'Ulysses': [0.00044682752457551384]})


#### Sorting the pangramiest book from the most to the least:

In [88]:
for w in sorted(pangrams_by_book, key=pangrams_by_book.get, reverse=True):
  print(w, pangrams_by_book[w])

Leaves of Grass [0.00233160621761658]
Ulysses [0.00044682752457551384]
Pride and Prejudice [0.00042005040604872583]
Moby Dick, or the Whale [0.00038387715930902113]
Wuthering Heights [0.00013795006207752792]
A Tale of Two Cities [0.00011811953697141507]
Oliver Twist or the Parish Boy's Progress [9.492168960607498e-05]
Frankenstein, or the Modern Prometheus [0.0]
Sense and Sensibility [0.0]
Great Expectations [0.0]


__Bonus:__ Print out the most efficient pangram and its author and book, as determined by fewest characters. (5 points)

In [91]:
def efficient_pangrams():
    BookNumber = [84, 98, 161, 730, 768, 1322, 1342, 1400, 2701, 4300]
    efficient = {}
    for booknumber in BookNumber:
        address = "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/books/"
        address = address + str(booknumber) + ".txt"
        sentences = []
        pansent = []
        efficient[booknumber] = []
        with open(address, "r") as file_handle:
            allthetext = file_handle.read()
            sentenceswithnewline = re.split("[\.\?\!][^a-zA-Z]", allthetext)
        for i in sentenceswithnewline:
            sentences.append(i.strip('\n'))
        for elements in sentences:
            if verify(str(elements)):
                pansent.append(str(elements))
        for each in pansent:
            letters = 0
            for i in range(len(each)):
                if each[i].isalpha():
                    letters += 1
            efficient[booknumber].append(letters)
        if efficient[booknumber]:
            efficient[booknumber] = min(efficient[booknumber])
        else:
            efficient.pop(booknumber)
    return(efficient)

In [92]:
d = efficient_pangrams()
print(d)

{98: 389, 730: 223, 768: 379, 1322: 417, 1342: 310, 2701: 391, 4300: 260}


In [103]:
most_efficient_Pangram_characters = min(d.values())
type(most_efficient_Pangram_characters)

int

In [106]:
BookNumber = [k for k,v in d.items() if v == most_efficient_Pangram_characters]
print(BookNumber[0])

730


In [120]:
pansent = []
sentences = []
with open(
    "/Users/Kianamon/Dropbox/DSCI511/KianaIsEditing/Assignments/assignment-1/data/books/730.txt", "r") as file_handle:
        allthetext = file_handle.read()
        sentenceswithnewline = re.split("[\.\?\!][^a-zA-Z]", allthetext)
        for i in sentenceswithnewline:
            sentences.append(i.strip('\n'))
        for elements in sentences:
            if verify(str(elements)):
                pansent.append(str(elements))

In [121]:
MostEfficientPangramInTheBooks = {}
MostEfficientPangramInTheBooks['Title'] = "Oliver Twist or the Parish Boy's Progress"
MostEfficientPangramInTheBooks['Author'] = 'Charles Dickens'
MostEfficientPangramInTheBooks['NumberOfChars'] = most_efficient_Pangram_characters
MostEfficientPangramInTheBooks['EfficientPangram'] = pansent[0].strip('\n') 
print(MostEfficientPangramInTheBooks)

{'Title': "Oliver Twist or the Parish Boy's Progress", 'Author': 'Charles Dickens', 'NumberOfChars': 223, 'EfficientPangram': 'At least half a\ndozen more were severally drawn forth from the same box, and surveyed\nwith equal pleasure; besides rings, brooches, bracelets, and other\narticles of jewellery, of such magnificent materials, and costly\nworkmanship, that Oliver had no idea, even of their names'}
